bt-agent: add new package

CarlosDerSeher opened the pull request:
Maintainer: me

Compile tested:
target-arm_cortex-a7+neon-vfpv4_musl_eabi
Model: Mikrotik hAP ac² (RBD52G-5HacD2HnD-TC)
```
# uname -a
# Linux 5.4.0-91-generic #102~18.04.1-Ubuntu SMP Thu Nov 11 14:46:36 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

Run tested:
Model: Mikrotik hAP ac² (RBD52G-5HacD2HnD-TC)
```
# uname -a
# Linux OpenWrt 5.10.80 #0 SMP Fri Nov 26 20:33:09 2021 armv7l GNU/Linux
```

tests done:
pairing using Android 7 Smartphone
connecting and playback using bluez-alsa configured as an a2dp sink

Description:
The Platform is also running bluez-alsa making it an A2DP sink. Pairing works with bluez-alsa service stopped. With bluez-alsa service started and configured as a2dp-sink it is also connectable and playback works flawless.

feckert commented on utils/bluez-tools/files/bt-agent.cfg:
```diff
@@ -0,0 +1,5 @@
+# format of pin file is: MAC_ADDR PIN
```
Can we convert this file generation to uci
```
config device
    option mac 'xx:xx:xx:xx:xx:xx'
    option pin '1234'
```

CarlosDerSeher replied:
probably yes but as I am not a 100% familiar with openWrt and this is my first contribution I am not sure how to proceed here. Do you have some manual or a package name where this is done so I have a guide how to do this?

CarlosDerSeher replied:
done

feckert commented on utils/bluez-tools/files/bt-agent.init:
```diff
@@ -0,0 +1,37 @@
+#!/bin/sh /etc/rc.common
```
Please use procd for this service init and remove all comments

CarlosDerSeher replied:
done

feckert commented on utils/bluez-tools/Makefile:
```diff
@@ -0,0 +1,50 @@
+PKG_SOURCE_PROTO := git
+PKG_SOURCE_URL := https://github.com/khvzak/bluez-tools.git
+PKG_SOURCE_VERSION := f653217
+PKG_BUILD_DIR := $(BUILD_DIR)/bluez-tools-$(PKG_VERSION)
```
Please remove the whitespaces before `:=` and after.
So that it looks consistent throughout the file

CarlosDerSeher replied:
done

feckert commented on utils/bluez-tools/files/bt-agent.init:
```diff
@@ -9,29 +11,33 @@ SERVICE_PID_FILE=/var/run/$SERVICE_NAME/pid
+	procd_set_param command "$DAEMON"
+	procd_append_param command "-c" "$AGENT_CAPABILITIES"
+	procd_append_param command "-p" "$AGENT_PIN_FILE"
+	procd_append_param command "&"
```
You don't have to add `&`. The procd takes care that the program runs as a service.

feckert commented on utils/bluez-tools/files/bt-agent.init:
```diff
@@ -9,29 +11,33 @@ SERVICE_PID_FILE=/var/run/$SERVICE_NAME/pid
-# Exit if the package is not installed
-[ -x "$DAEMON" ] || exit 0
-
 AGENT_PIN_FILE=/etc/bt-agent.cfg
```
Pid files are located in `/var/run/`, so please change this to `/var/run/bt-agent.pid`

CarlosDerSeher replied:
this is the pin file needed by bt-agent not a pid file

feckert replied:
Ok, I read that wrong.
But since we are on an embedded system, the write accesses to the NAND should be reduced.
Therefore it is a must that we move the generated file to /tmp and not keep it in /etc.

CarlosDerSeher replied:
Alright that seems like a logical thing to do. I will update this and the other stuff as soon as I got some spare time.

feckert commented on utils/bluez-tools/files/bt-agent.init:
```diff
@@ -9,29 +11,33 @@ SERVICE_PID_FILE=/var/run/$SERVICE_NAME/pid
 DESC="Bluetooth pairing agent"
```
This is not needed and used?

CarlosDerSeher replied:
done

feckert commented on utils/bluez-tools/files/bt-agent.init:
```diff
@@ -9,29 +11,33 @@ SERVICE_PID_FILE=/var/run/$SERVICE_NAME/pid
-	#killall $SERVICE_NAME
-	# Many daemons don't delete their pidfiles when they exit.
-	rm -f $SERVICE_PID_FILE
+	config_load bt-agent
```
I am not a fan of rewriting the callback for uci.
Please use the [config_foreach](
https://github.com/openwrt/packages/search?q=config_foreach) pattern for uci.

CarlosDerSeher replied:
done

feckert commented on utils/bluez-tools/Makefile:
```diff
@@ -38,7 +38,8 @@ endef
 	$(INSTALL_DIR) $(1)/etc/init.d
 	$(INSTALL_BIN) ./files/bt-agent.init $(1)/etc/init.d/bt-agent
-	$(CP) ./files/bt-agent.cfg $(1)/etc/bt-agent.cfg
+	$(INSTALL_DIR) $(1)/etc/config
```
You also need to place the uci config in `/etc/config `without the `cfg` extension.
Also add a `define Package/bluez-tools/conffiles` definition to your makefile https://github.com/openwrt/packages/search?q=conffiles, so this gets save during sysupgrade.

CarlosDerSeher replied:
done

feckert commented on utils/bluez-tools/Makefile:
```diff
@@ -1,23 +1,23 @@
-#
+ #
```
remove `space` before `#`

CarlosDerSeher replied:
done

CarlosDerSeher commented:
Hi there, just wondering if there is any progress regarding my pull request?

feckert commented:
@CarlosDerSeher Well, it's still a draft from your side, if this not the case, then you should have removed it. Also, it wouldn't have been bad if you had highlighted me with @feckert. I'll pull the pullrequest again and see if everything fits. But I haven't seen anything conspicuous now. you have implemented my comment.

CarlosDerSeher commented:
@feckert sorry about the confusion, I didn't realize the draft state of my pull request. I changed the status now. Thank you for your time and patience, I am not that experienced with using GitHub yet

feckert commented:
I looked at it again. There are my remarks.
* I have now looked at [debian](https://packages.debian.org/bullseye/bluez-tools), there is a version 2.0 of this program. They use the same source. Could we also set this do version 2.0? Or explain me why not.
* Can you please merge the commits to one, so we have only one commit `bluez-tools: Add package bluezetools` the other commits are not of interest, as this is a new tool.
* Can you also please change the file `/etc/config/bt-agent` from 755 to 644 with chmod. This is not an executable.
* I also saw that the init script is with a '-'. I'm not sure if this is allowed. I have only seen a '_' as a separator for init script in openwrt. I would suggest we name the init script `btagent`

CarlosDerSeher commented:
I got the version from here https://code.google.com/archive/p/bluez-tools/
I am not exactly sure how up to date this site is. I didn't find any other information regarding a version. I definitely could change this to the debian version. Should we download the source as a tar from there?
I'll implement the other remarks as soon as I can

CarlosDerSeher commented:
@feckert Ok I believe I am done now. The only thing I am not sure about is the Version string 2.0~20201025.**0**.f653217**-4**
 am not exactly sure about the parts written bold. I just copied the version from your link to Debian and changed to the commit-hash with its date I use in the Makefile

feckert commented:
Looks good so far. I'm not sure about the version either! The version number of debian is not an official number of the project. There is no git tag in the [official repo](https://github.com/khvzak/bluez-tools).

Because of useing the git tree directly, I would suggest, that we use the last commit date in the master branch as a version number. https://github.com/khvzak/bluez-tools/commits/master This is `PKG_SOURCE_DATE:=2020-10-25`

Please also use the full commit hash number for `PKG_SOURCE_VERSION:=f65321736475429316f07ee94ec0deac8e46ec4a`

After we have downloaded the whole thing via git we should create the hash for the tar.gz and add it to the Makefile `PKG_MIRROR_HASH=:xxxxxx` also.Then you don't always have to connect to github to build the package, but can use the tar.gz in your dl directory

CarlosDerSeher commented:
@feckert Ok I add the changes as requested. For the Version I chose 20201025.f653217 a combination of git commit date and commit-hash, is this ok?

feckert commented on utils/bluez-tools/Makefile:
```diff
@@ -0,0 +1,54 @@
+
+PKG_NAME:=bluez-tools
+PKG_VERSION:=20201025.f653217
+PKG_RELEASE:=$(PKG_SOURCE_VERSION)
```
Sorry but I still don't agree.
This should be `PKG_RELEASE:=1` and should be update if you changes the Makefile or add a patch to change the source.
You could also use `PKG_RELEASE:=$(AUTORELEASE)` Then the commit history is used to check the last **PKG_VERSION** update, and increments the number for each commit automaticaly.

feckert commented on utils/bluez-tools/Makefile:
```diff
@@ -0,0 +1,54 @@
+PKG_VERSION:=20201025.f653217
+PKG_RELEASE:=$(PKG_SOURCE_VERSION)
+
+PKG_MAINTAINER:=Alex Orlenko <none>, Karl Osterseher <karli_o@gmx.at>
```
I think we also need a valid email address for the maintainer **Alex Orlenko**
The maintainer is needed or should be indicated who feels responsible for the package in OpenWrt and can therefore be contacted if there are problems with the package. Therefore a valid email address should be entered here, or you remove him.

feckert reviewed (changes requested):
Everything else looks good. I have already compiled the package

CarlosDerSeher commented:
@feckert ok as I am the maintainer of the openWrt package bluez-tools I removed Alex Orlenko. I thought because he owns https://github.com/khvzak/bluez-tools he also needs to be there. Also changed to PKG_RELEASE:=$(AUTORELEASE)

feckert commented:
Then let's try
Thanks for your contribution
